### PR TITLE
Added faddeeva_w function

### PR DESCRIPTION
**Pull request by fp4code**
"w", the mother of all Faddeva functions, can be useful. This function is named wofz in python/scipy (Cf. http://ab-initio.mit.edu/wiki/index.php/Faddeeva_Package). It is called faddeeva_w in this PR.

### CODE REVIEW

**Comment by stevengj** [comment]
I would just call it `faddeeva(z)` if we add this; `faddeeva_w` seems unnecessarily verbose since there is no other "Faddeeva" function in common parlance.

I'm a little skeptical that we need to add it explicitly, however, since it is basically redundant with `erfcx`: `fadeeva(z) = erfcx(-im*z)`.

**Comment by fp4code** [comment]
We can think to implement raw original C-functions with relerr argument as faddeeva.z(z, relerr), faddeeva.erf(z, relerr), etc.
Hence the choice of faddeeva_w instead of faddeeva. But capitalized form Faddeeva.z(z, relerr), Faddeeva.erf(z, relerr) etc. can be choosen for raw functions and faddeeva(z) for high precision function.

About the redundancy, there is a modest gain with fadeeva_w(z) over erfcx(-im*z): between 5% and 7% on timing tests.
```
julia> w(z)=SpecialFunctions.erfcx(-im*z)
w (generic function with 1 method)

julia> @btime w(2.1)
  30.296 ns (0 allocations: 0 bytes)
0.012155178329914935 + 0.3180730926015565im

julia> @btime SpecialFunctions.faddeeva_w(2.1)
  28.904 ns (0 allocations: 0 bytes)
0.012155178329914935 + 0.3180730926015565im

julia> @btime w(0.1)
  21.997 ns (0 allocations: 0 bytes)
0.9900498337491681 + 0.11208866436449542im

julia> @btime SpecialFunctions.faddeeva_w(0.1)
  20.607 ns (0 allocations: 0 bytes)
0.9900498337491681 + 0.11208866436449542im
```

**Comment by stevengj** [comment]
> About the redundancy, there is a modest gain with fadeeva_w(z) over erfcx(-im*z): between 5% and 7% on timing tests.

I can believe that, except:

* You can probably make it a bit faster by using `erfcx(Complex(imag(z), -real(z)))` if this 5% difference matters.

* In real applications, the fadeeva function is typically combined with other calculations, e.g. to compute Voigt profiles, and the `-im*z` calculation can often be combined with the other calculations at no cost.

**Comment by stevengj** [comment]
(The relationship to `erfcx` is precisely why I didn't wrap this function in my initial work to connect this to Julia: JuliaLang/julia#1808.  Actually, putting this in Julia was my very first pull request to Julia JuliaLang/julia#1799.  That may not be an argument in favor of my design choices back then 😉.)

That being said, if someone googles a formula for e.g. Voigt profiles and finds something in terms of w(z), it is a bit friendlier if we provide a `faddeeva(z)` function directly rather than requiring them to understand the relationship to `erfcx`.

**Comment by fp4code** [comment]
OK, this PR can be withdrawn. Maybe the documentation can be completed to say that _w_ function (as python scilab.special.wofz) can be defined by the user as erfcx(Complex(imag(z), -real(z))). Just to educate   faddeeva newbies.

**Comment by stevengj** [comment]
It's probably easier to just add a `faddeeva(z)` function than to try and get users to use `erfcx` for this.

**Comment by fp4code** [comment]
There was a crosspost between us. As you are inclined to keep this PR, the function is renamed now.

File: src/erf.jl [hunk 1]
```diff
@@ -38,9 +38,9 @@ for ff in (:erfcx, :erfi, (:dawson, :Dawson))
     end
 end
 
-faddeeva_w(x::Float64) = faddeeva_w(Complex{Float64}(x))
-faddeeva_w(x::Float32) = faddeeva_w(Complex{Float32}(x))
-faddeeva_w(x::Integer) = faddeeva_w(float(x))
+faddeeva(x::Float64) = faddeeva(Complex{Float64}(x))
+faddeeva(x::Float32) = faddeeva(Complex{Float32}(x))
+faddeeva(x::Integer) = faddeeva(float(x))
```
Comment (stevengj) [inline src/erf.jl]: Probably just
```jl
faddeeva(x::Real) = faddeeva(complex(x))
```
?

File: docs/src/index.md [hunk 2]
```diff
@@ -12,6 +12,7 @@ libraries.
 | [`erfcinv(x)`](@ref SpecialFunctions.erfcinv)                 | inverse function to [`erfc()`](@ref SpecialFunctions.erfc)                                                                                                       |
 | [`erfi(x)`](@ref SpecialFunctions.erfi)                       | imaginary error function defined as `-im * erf(x * im)`, where `im` is the imaginary unit                                                                       |
 | [`erfcx(x)`](@ref SpecialFunctions.erfcx)                     | scaled complementary error function, i.e. accurate `exp(x^2) * erfc(x)` for large `x`                                                                           |
+| [`faddeeva(x)`](@ref SpecialFunctions.faddeeva)               | faddeva function `exp(-x^2) * erfc(-ix)`                                                                           |
```
Comment (stevengj) [inline docs/src/index.md]: `= erfcx(-ix)`

File: src/erf.jl [hunk 3]
```diff
@@ -73,6 +76,17 @@ Compute the Dawson function (scaled imaginary error function) of `x`, defined by
 """
 dawson
 
+"""
+    faddeeva(z)
+
+Compute the Faddeeva function of complex `z`, defined by
+``e^{-z^2} \\operatorname{erfc}(-iz)``.
+Note that this function, also named `w` (original Faddeeva package) or `wofz` (Scilab package)
+is equivalent to``\\operatorname{erfcx}(-ix)``.
```
Comment (stevengj) [inline src/erf.jl]: z, not x

File: src/SpecialFunctions.jl [hunk 4]
```diff
@@ -21,7 +21,7 @@ if isdefined(Base, :airyai) && VERSION < v"0.7.0-DEV.986" #22763
                  besselh, besselhx, besseli, besselix, besselj, besselj0, besselj1,
                  besseljx, besselk, besselkx, bessely, bessely0, bessely1, besselyx,
                  hankelh1, hankelh1x, hankelh2, hankelh2x,
-                 dawson, erf, erfc, erfcinv, erfcx, erfi, erfinv,
+                 dawson, erf, erfc, erfcinv, erfcx, erfi, erfinv, faddeeva,
```
Comment (stevengj) [inline src/SpecialFunctions.jl]: You can’t import this from Base

Comment (fp4code) [reply src/SpecialFunctions.jl]: Indeed, do you agree to do nothing for VERSION < v"0.7.0-DEV.986" ?

File: src/erf.jl (hunk 3 above)
Comment (fp4code) [reply src/erf.jl]: Oops, I will do that.

File: src/SpecialFunctions.jl (hunk 4 above)
Comment (stevengj) [reply src/SpecialFunctions.jl]: The right thing is to `export faddeeva` in these old versions, not to do nothing.

File: src/SpecialFunctions.jl [hunk 5]
```diff
@@ -21,7 +21,7 @@ if isdefined(Base, :airyai) && VERSION < v"0.7.0-DEV.986" #22763
                  besselh, besselhx, besseli, besselix, besselj, besselj0, besselj1,
                  besseljx, besselk, besselkx, bessely, bessely0, bessely1, besselyx,
                  hankelh1, hankelh1x, hankelh2, hankelh2x,
-                 dawson, erf, erfc, erfcinv, erfcx, erfi, erfinv, faddeeva,
```
Comment (stevengj) [inline src/SpecialFunctions.jl]: You still need to export `faddeeva`

Comment (fp4code) [reply src/SpecialFunctions.jl]: Thanks to have patience to a newbie, Steven!

File: src/erf.jl [hunk 6]
```diff
@@ -73,6 +76,17 @@ Compute the Dawson function (scaled imaginary error function) of `x`, defined by
 """
 dawson
 
+"""
+    faddeeva(z)
+
+Compute the Faddeeva function of complex `z`, defined by
+``e^{-z^2} \\operatorname{erfc}(-iz)``.
+Note that this function, also named `w` (original Faddeeva package) or `wofz` (Scilab package)
```
Comment (stevengj) [inline src/erf.jl]: missing a comma at the end of this line.

Comment (fp4code) [reply src/erf.jl]: yep, done

File: src/erf.jl [hunk 7]
```diff
@@ -4,7 +4,7 @@ using Base.Math: @horner, libm
 using Base.MPFR: ROUNDING_MODE
 
 for f in (:erf, :erfc)
-    @eval begin
+    @eval beginhttps://github.com/discourse/discourse/tree/master/config/locales
```
Comment (stevengj) [inline src/erf.jl]: whoops?

Comment (ararslan) [reply src/erf.jl]: Seems like valid syntax to me :trollface:

**Comment by stevengj** [comment]
This PR reminds me of another thing I don't like about the Faddeeva function: unlike `erfcx`, it doesn't give real outputs for real inputs, which makes `erfcx` a much nicer way to present essentially the same function.  (`faddeeva` gives real outputs for purely imaginary inputs.)

**Comment by fp4code** [comment]
You are right, but you need to take into account the "marketing" aspect for non specialists of thoses functions. Reading without high concentration the page http://ab-initio.mit.edu/wiki/index.php/Faddeeva_Package, Julia seems less complete than other language libraries.

File: src/erf.jl (hunk 7 above)
Comment (fp4code) [reply src/erf.jl]: Sorry. For few months, my linux laptop is pasting unexpectedly the buffer from time to time. I've just made some experiments now. Touchpad tapping is disable for years, but right top angle touchpad tapping is strangely enabled now. I will remove that.

**Comment by ViralBShah** [comment]
@stevengj Should we resolve conflicts and merge this?

**Comment by stevengj** [comment]
Resolved the conflicts.

**Comment by cossio** [comment]
Merge?

**Comment by stevengj** [comment]
I guess it's fine, though I'm not super enthusiastic as it's mostly just a "marketing" function.  But at least the implementation is short and the name is unlikely to conflict with anything else.

**Comment by simonbyrne** [comment]
needs another update.

**Comment by ViralBShah** [comment]
It seems like this bitrotted away. Since we had got it close enough, it would be nice to get this in.

**Review by ararslan — APPROVED** [review]

**Comment by ararslan** [comment]
This PR will be 5 years old next month! Steven and I have formally reviewed and approved and other commenters have suggested support, so let's merge it before we forget and have to rebase it again in a couple of years. Thanks for the contribution and for your patience, @fp4code!

**Pull request merged**